packages/salesforcedx-vscode-core/src/commands/util/commandLog.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { CommandEvent, CommandEventStream, CommandEventType } from '@salesforce/salesforcedx-utils-vscode';
+import { WorkspaceContext } from '../../context';
+import { CommandLogEntry } from './commandLogEntry';
+
+export const getCommandLog = (commandIdFilter?: string, exitCodeFilter?: number): CommandLogEntry[] => {
+  return CommandLog.getInstance().getCommandLog(commandIdFilter, exitCodeFilter);
+};
+
+export const getLastCommandLogEntry = (commandIdFilter?: string, exitCodeFilter?: number): CommandLogEntry | undefined => {
+  const commandLog = CommandLog.getInstance().getCommandLog(commandIdFilter, exitCodeFilter);
+  return commandLog.length > 0 ? commandLog[commandLog.length - 1] : undefined;
+};
+
+export class CommandLog {
+  private static instance: CommandLog;
+  private static STORAGE_KEY = 'COMMAND_LOG';
+  private static MAX_LOG_ENTRIES = 1000;
+
+  private commandLogEntries: CommandLogEntry[] = [];
+  private inProgressCommands: Map<string, number> = new Map();
+  private inProgressData: any = {};
+
+  private constructor() {
+    this.loadCommandLog();
+  }
+
+  public static getInstance(): CommandLog {
+    if (!CommandLog.instance) {
+      CommandLog.instance = new CommandLog();
+    }
+    return CommandLog.instance;
+  }
+
+  public initialize() {
+    CommandEventStream.getInstance().onCommandEvent(this.processCommandEvent.bind(this));
+  }
+
+  public async logCommand(commandId: string, duration: number) {
+    const entry = {
+      commandId,
+      timestamp: Date.now(),
+      duration,
+      exitCode: this.inProgressData.exitCode,
+      error: this.inProgressData.error,
+      data: this.inProgressData.data
+    };
+    this.commandLogEntries.push(entry);
+    if (this.commandLogEntries.length > CommandLog.MAX_LOG_ENTRIES) {
+      this.commandLogEntries.shift();
+    }
+    await this.updateCommandLog();
+  }
+
+  public getCommandLog(commandIdFilter?: string, exitCodeFilter?: number): CommandLogEntry[] {
```

Comment:
is commandIdFilter purposely optional?   
I guess it comes with the benefit that I could get all command with -1 exit code, but I'm more concerned with code that would accidentally pass undefined as the commandId then react as if they got back command that's weren't expected. 
```
let startCmd = undefined;
let failedStartCommands = getCommandLog(startCmd, -1) 
console.log('All failed start commands', failedStartCommands);
```

Reply:
Yeah I wanted to leave open the option for clients to get multiple types of commands. But parameter order makes it seem like it shouldn't be `undefined`. It seems to me like an argument for:

> Yeah it's really not necessary. It's easy enough for a client of this to find the last one. I'm thinking about just having getCommandLog and an options parameter that has the commandIdFilter and exitCodeFilter and maybe a timestamp sort order as well. Overkill?